File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/mediator/concrete_mediator.dart
import 'colleague.dart';
import 'concrete_colleague_a.dart';
import 'concrete_colleague_b.dart';
import 'concrete_colleague_c.dart';
import 'concrete_colleague_d.dart';
import 'mediator.dart';
import 'types.dart';

class ConcreteMediator implements Mediator {
  final void Function(String line) onLog;
  final SearchBox A;
  final ResultsList B;
  final CategoryFilter C;
  final ClearButton D;
  final List<Item> dataset;

  ConcreteMediator(
    this.A,
    this.B,
    this.C,
    this.D,
    this.dataset, {
    void Function(String line)? onLog,
  }) : onLog = onLog ?? ((_) {}) {
    A.setMediator(this);
    B.setMediator(this);
    C.setMediator(this);
    D.setMediator(this);
  }

  @override
  void notify(Colleague sender, String event, [Object? payload]) {
    if (sender == A) {
      _reactOnA(event, payload as String? ?? '');
    } else if (sender == B) {
      _reactOnB(event, (payload as int?) ?? 0);
    } else if (sender == C) {
      _reactOnC(event, payload as String? ?? 'all');
    } else if (sender == D) {
      _reactOnD(event);
    }
  }

  void _reactOnA(String _, String q) {
    onLog('A → query="$q"');
    _refreshResults();
  }

  void _reactOnB(String _, int n) {
    onLog('B → rendered $n items');
  }

  void _reactOnC(String _, String cat) {
    onLog('C → category=$cat');
    _refreshResults();
  }

  void _reactOnD(String _) {
    onLog('D → clear all');
    A.value = '';
    C.value = 'all';
    _refreshResults();
  }

  void _refreshResults() {
    final q = A.value.trim().toLowerCase();
    final cat = C.value;
    final filtered = dataset
        .where(
          (it) =>
              (cat == 'all' || it.category == cat) &&
              (q.isEmpty || it.name.toLowerCase().contains(q)),
        )
        .toList(growable: false);
    B.operationB(filtered);
  }
}

File: implementations/react-native/design-patterns-rn/src/patterns/behavioral/mediator/concrete-mediator.ts
import type { Colleague } from './colleague';
import type { SearchBox } from './concrete-colleague-a';
import type { ResultsList } from './concrete-colleague-b';
import type { CategoryFilter } from './concrete-colleague-c';
import type { ClearButton } from './concrete-colleague-d';
import type { Mediator } from './mediator';
import type { Item } from './types/type';

export class ConcreteMediator implements Mediator {
  private onLog: (line: string) => void;
  private A: SearchBox;
  private B: ResultsList;
  private C: CategoryFilter;
  private D: ClearButton;
  private dataset: Item[];
  constructor(
    A: SearchBox,
    B: ResultsList,
    C: CategoryFilter,
    D: ClearButton,
    dataset: Item[],
    onLog?: (line: string) => void
  ) {
    this.A = A;
    this.B = B;
    this.C = C;
    this.D = D;
    this.dataset = dataset;
    this.onLog = onLog ?? (() => {});
    A.setMediator(this); B.setMediator(this); C.setMediator(this); D.setMediator(this);
  }

  notify(sender: unknown, event: string, payload?: unknown): void {
    if (sender === this.A) this.reactOnA(event, payload as string);
    else if (sender === this.B) this.reactOnB(event, payload as number);
    else if (sender === this.C) this.reactOnC(event, payload as any);
    else if (sender === this.D) this.reactOnD(event);
  }

  private reactOnA(_: string, q: string) {
    this.onLog(`A → query="${q}"`);
    this.refreshResults();
  }
  private reactOnB(_: string, n: number) {
    this.onLog(`B → rendered ${n} items`);
  }
  private reactOnC(_: string, cat: 'all' | Item['category']) {
    this.onLog(`C → category=${cat}`);
    this.refreshResults();
  }
  private reactOnD(_: string) {
    this.onLog('D → clear all');
    this.A.value = '';
    this.C.value = 'all';
    this.refreshResults();
  }

  private refreshResults() {
    const q = this.A.value.trim().toLowerCase();
    const cat = this.C.value;
    const filtered = this.dataset.filter(it =>
      (cat === 'all' || it.category === cat) &&
      (q === '' || it.name.toLowerCase().includes(q))
    );
    this.B.operationB(filtered);
  }
}
